include/morphio/section_iterators.hpp

```diff
@@ -0,0 +1,252 @@
+#pragma once
+
+#include <algorithm> // std::copy
+#include <deque> // std::deque
+#include <iterator> // std::back_inserter / std::front_inserter
+#include <memory> // std::shared_ptr
+#include <vector> // std::vector
+
+#include <morphio/types.h>
+#include <morphio/exceptions.h>
+
+namespace detail {
+
+template<typename SectionT, typename MorphologyT>
+std::vector<SectionT> getChildren(const MorphologyT& morphology)
+{
+    return morphology.rootSections();
+}
+
+template<typename SectionT>
+std::vector<SectionT> getChildren(const SectionT& section)
+{
+    return section.children();
+}
+
+template<typename SectionT>
+std::vector<std::shared_ptr<SectionT>> getChildren(const std::shared_ptr<SectionT>& section)
+{
+    return section->children();
+}
+
+template<typename SectionT>
+bool isRoot(const SectionT& current)
+{
+    return current.isRoot();
+}
+
+template<typename SectionT>
+bool isRoot(const std::shared_ptr<SectionT>& current)
+{
+    return current->isRoot();
+}
+
+template<typename SectionT>
+SectionT getParent(const SectionT& current)
+{
+    return current.parent();
+}
+
+template<typename SectionT>
+std::shared_ptr<SectionT> getParent(const std::shared_ptr<SectionT>& current)
+{
+    return current->parent();
+}
+} // namespace detail
+
+namespace morphio {
+
+template<typename SectionT, typename MorphologyT>
+class breadth_iterator_t {
+public:
+    breadth_iterator_t() = default;
+
+    explicit breadth_iterator_t(const SectionT& section)
+    {
+        deque_.push_front(section);
+    }
+
+    explicit breadth_iterator_t(const MorphologyT& morphology)
+    {
+        const auto& children = detail::getChildren<SectionT, MorphologyT>(morphology);
+        std::copy(children.begin(), children.end(), std::back_inserter(deque_));
+    }
+
+    breadth_iterator_t(const breadth_iterator_t& other)
+    : deque_(other.deque_)
+    {}
+
+    SectionT operator*() const
+    {
+        return deque_.front();
+    }
+
+    breadth_iterator_t operator++()
+    {
+        if (deque_.empty()) {
+            LBTHROW(MissingParentError("XXXXXXXXXXXXXXXXXXX"));
```

Comment:
It's not really a missing parent error no ?

Reply:
Nope, wanted to ask: which exception do you prefer?

Reply:
The basic `MorphioError` or a new `IteratorError` one ?